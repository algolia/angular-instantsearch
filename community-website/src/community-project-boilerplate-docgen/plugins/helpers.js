const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
