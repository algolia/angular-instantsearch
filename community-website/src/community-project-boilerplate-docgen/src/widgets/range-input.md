---
title: Widgets - Range Input
layout: widget.pug
---

# Range Input

## Description

Numeric range widget allows a user to select a numeric range using a minimum and maximum input.

## Options

| Attribute        | Type       | Description
| -                | -          | -
| `attribute`      | `string`   | Name of the attribute for faceting
| `min?`           | `number`   | Minimal value, default to automatically computed
| `max?`           | `number`   | Maximal value, default to automatically computed
| `precision?`     | `number`   | Number of digits after decimal point to use

## Code example

```html
<ais-range-input
  attribute="price"
  [min]="10"
  [max]="10"
  [precision]="2"
>
</ais-range-input>
```
