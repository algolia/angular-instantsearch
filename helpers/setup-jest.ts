import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
