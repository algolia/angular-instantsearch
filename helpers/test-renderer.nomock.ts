import { Component, ViewChild } from '@angular/core';
import { TestBed } from '@angular/core/testing';

import { NgAisInstantSearchModule } from '../src/instantsearch/instantsearch.module';

// same as test-renderer, but without mocking InstantSearch
// In the future, we want to migrate all tests to this

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations,
  methods = {},
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
  methods?: {};
}) {
  return function(state?: {}, isFirstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined,
        methods,
      },
      isFirstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state,
    methods,
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
    methods?: {};
  },
  isFirstRender = false
) {
  @Component({
    template: `
      <ais-instantsearch [config]="config">
        ${template}
      </ais-instantsearch>
    `,
  })
  class TestContainer {
    config = {
      searchClient: {
        search(requests) {
          return Promise.resolve({
            results: requests.map(() => ({ hits: [] })),
          });
        },
      },
      indexName: 'instant_search',
    };
    @ViewChild(TestedWidget) testedWidget;
    constructor() {
      Object.keys(methods).forEach(methodName => {
        this[methodName] = methods[methodName];
      });
    }
  }

  TestBed.configureCompiler({
    preserveWhitespaces: false,
  } as any).configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || []),
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])],
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, isFirstRender);
    fixture.detectChanges();
  }

  return fixture;
}
