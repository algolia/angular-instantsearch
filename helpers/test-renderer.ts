import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
