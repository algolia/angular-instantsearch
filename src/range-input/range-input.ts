import { Inject, Component, Input, forwardRef } from '@angular/core';

import { connectRange } from 'instantsearch.js/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgAisInstantSearch } from '../instantsearch/instantsearch';
import { parseNumberInput, noop } from '../utils';

export type NumericRangeState = {
  start: number[];
  range: { min?: number; max?: number };
  refine: Function;
};

@Component({
  selector: 'ais-range-input',
  template: `
    <div [class]="cx()">
      <form
        [class]="cx('form')"
        (submit)="handleSubmit($event)"
        novalidate
      >
        <label [class]="cx('label')">
          <span [class]="cx('currency')">{{currency}}</span>
          <input
            [class]="cx('input', 'min')"
            type="number"
            [min]="state.range.min"
            [max]="state.range.max"
            [placeholder]="state.range.min"
            [value]="minInputValue"
            [step]="step"
            (change)="handleChange($event, 'min')"
          />
        </label>

        <span [class]="cx('separator')">{{separator}}</span>

        <label [class]="cx('label')">
          <span [class]="cx('currency')">{{currency}}</span>
          <input
            [class]="cx('input', 'max')"
            type="number"
            [min]="state.range.min"
            [max]="state.range.max"
            [placeholder]="state.range.max"
            [value]="maxInputValue"
            [step]="step"
            (change)="handleChange($event, 'max')"
          />
        </label>

        <button
          [class]="cx('submit')"
          (click)="handleSubmit($event)"
        >
          {{submitLabel}}
        </button>
      </form>
    </div>
  `,
})
export class NgAisRangeInput extends BaseWidget {
  // rendering options
  @Input() public currency: string = '$';
  @Input() public separator: string = 'to';
  @Input() public submitLabel: string = 'Go';

  // instance options
  @Input() public attribute: string;
  @Input() public min?: number;
  @Input() public max?: number;
  @Input() public precision?: number;

  // inner state
  public minInputValue?: number | string = '';
  public maxInputValue?: number | string = '';

  get step() {
    const precision = parseNumberInput(this.precision) || 2;
    return 1 / Math.pow(10, precision);
  }

  public state: NumericRangeState = {
    range: { min: undefined, max: undefined },
    refine: noop,
    start: [0, 0],
  };

  constructor(
    @Inject(forwardRef(() => NgAisInstantSearch))
    public instantSearchParent: any
  ) {
    super('RangeInput');
  }

  public ngOnInit() {
    this.createWidget(connectRange, {
      attribute: this.attribute,
      max: parseNumberInput(this.max),
      min: parseNumberInput(this.min),
      precision: parseNumberInput(this.precision),
    });

    super.ngOnInit();
  }

  public handleChange(event: any, type: string) {
    const value = parseNumberInput(event.target.value);

    if (type === 'min') {
      this.minInputValue = value;
    } else {
      this.maxInputValue = value;
    }
  }

  public handleSubmit(event: MouseEvent | KeyboardEvent) {
    event.preventDefault();
    this.state.refine([this.minInputValue, this.maxInputValue]);
  }
}
