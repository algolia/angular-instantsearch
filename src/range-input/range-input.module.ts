import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";

import { NgAisRangeInput } from "./range-input";

@NgModule({
  declarations: [NgAisRangeInput],
  entryComponents: [NgAisRangeInput],
  exports: [NgAisRangeInput],
  imports: [CommonModule]
})
export class NgAisRangeInputModule {}
