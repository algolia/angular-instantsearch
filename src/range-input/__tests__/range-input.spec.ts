import { createRenderer } from '../../../helpers/test-renderer';
import { NgAisRangeInput } from '../range-input';

const defaultState = {
  range: { min: 0, max: 100 },
  start: [0, 100],
  refine: jest.fn(),
};

const render = createRenderer({
  defaultState,
  template: '<ais-range-input></ais-range-input>',
  TestedWidget: NgAisRangeInput,
});

describe('RangeInput', () => {
  it('renders markup without state', () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it('renders markup with state', () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it('should update `min/max InputValue`', () => {
    const fixture = render({});

    const [
      minInput,
      maxInput,
    ] = fixture.debugElement.nativeElement.querySelectorAll('input');

    minInput.value = '20';
    minInput.dispatchEvent(new Event('change'));

    maxInput.value = '50';
    maxInput.dispatchEvent(new Event('change'));

    expect(fixture.componentInstance.testedWidget.minInputValue).toBe(20);
    expect(fixture.componentInstance.testedWidget.maxInputValue).toBe(50);
  });

  it('should call renfine when submitting form', () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const [
      minInput,
      maxInput,
    ] = fixture.debugElement.nativeElement.querySelectorAll('input');

    minInput.value = '20';
    minInput.dispatchEvent(new Event('change'));

    maxInput.value = '50';
    maxInput.dispatchEvent(new Event('change'));

    expect(fixture.componentInstance.testedWidget.minInputValue).toBe(20);
    expect(fixture.componentInstance.testedWidget.maxInputValue).toBe(50);

    const submitBtn = fixture.debugElement.nativeElement.querySelector(
      'button'
    );
    submitBtn.click();

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith([20, 50]);
  });
});
