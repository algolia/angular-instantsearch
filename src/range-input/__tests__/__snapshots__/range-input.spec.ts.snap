// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeInput renders markup with state 1`] = `
<ng-ais-range-input
  currency={[Function String]}
  cx={[Function Function]}
  maxInputValue=""
  minInputValue=""
  precision={[Function Number]}
  searchInstance={[Function NgAisInstance]}
  separator={[Function String]}
  state={[Function Object]}
  submitLabel={[Function String]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-RangeInput"
  >
    
        
    <form
      class="ais-RangeInput-form"
      novalidate=""
    >
      
          
      <label
        class="ais-RangeInput-label"
      >
        
            
        <span
          class="ais-RangeInput-currency"
        >
          $
        </span>
        
            
        <input
          class="ais-RangeInput-input--min"
          max="100"
          min="0"
          placeholder="0"
          step="0.01"
          type="number"
        />
        
          
      </label>
      
  
          
      <span
        class="ais-RangeInput-separator"
      >
        to
      </span>
      
  
          
      <label
        class="ais-RangeInput-label"
      >
        
            
        <span
          class="ais-RangeInput-currency"
        >
          $
        </span>
        
            
        <input
          class="ais-RangeInput-input--max"
          max="100"
          min="0"
          placeholder="100"
          step="0.01"
          type="number"
        />
        
          
      </label>
      
  
          
      <button
        class="ais-RangeInput-submit"
      >
        
            Go
          
      </button>
      
        
    </form>
    
      
  </div>
    
</ng-ais-range-input>
`;

exports[`RangeInput renders markup without state 1`] = `
<ng-ais-range-input
  currency={[Function String]}
  cx={[Function Function]}
  maxInputValue=""
  minInputValue=""
  precision={[Function Number]}
  searchInstance={[Function NgAisInstance]}
  separator={[Function String]}
  state={[Function Object]}
  submitLabel={[Function String]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-RangeInput"
  >
    
        
    <form
      class="ais-RangeInput-form"
      novalidate=""
    >
      
          
      <label
        class="ais-RangeInput-label"
      >
        
            
        <span
          class="ais-RangeInput-currency"
        >
          $
        </span>
        
            
        <input
          class="ais-RangeInput-input--min"
          max="undefined"
          min="undefined"
          placeholder="undefined"
          step="0.01"
          type="number"
        />
        
          
      </label>
      
  
          
      <span
        class="ais-RangeInput-separator"
      >
        to
      </span>
      
  
          
      <label
        class="ais-RangeInput-label"
      >
        
            
        <span
          class="ais-RangeInput-currency"
        >
          $
        </span>
        
            
        <input
          class="ais-RangeInput-input--max"
          max="undefined"
          min="undefined"
          placeholder="undefined"
          step="0.01"
          type="number"
        />
        
          
      </label>
      
  
          
      <button
        class="ais-RangeInput-submit"
      >
        
            Go
          
      </button>
      
        
    </form>
    
      
  </div>
    
</ng-ais-range-input>
`;
